File: frontend/src/services/terraform.ts
import { api } from './api'

export interface TerraformWorkspace {
  id: string
  name: string
  description?: string | null
  source_type: string
  repository_url?: string | null
  branch?: string | null
  root_path?: string | null
  provider?: string | null
  environment?: string | null
  owner?: string | null
  last_scan_id?: string | null
  created_at: string
  updated_at: string
}

export interface TerraformFinding {
  id: string
  workspace_id: string
  scan_id?: string | null
  resource_id?: string | null
  title: string
  description: string
  impact?: string | null
  severity: string
  category: string
  status: string
  scanner: string
  rule_id?: string | null
  resource_address?: string | null
  resource_type?: string | null
  file_path?: string | null
  line_number?: number | null
  remediation?: string | null
  ai_explanation?: string | null
  confidence_score?: number | null
  created_at: string
}

export interface TerraformDrift {
  id: string
  workspace_id: string
  scan_id?: string | null
  resource_id?: string | null
  resource_address: string
  resource_type?: string | null
  attribute_path: string
  desired_value?: unknown
  actual_value?: unknown
  drift_type: string
  severity: string
  status: string
  description: string
  remediation?: string | null
  confidence_score?: number | null
  created_at: string
}

export interface TerraformScan {
  id: string
  workspace_id?: string | null
  scan_name: string
  source_type: string
  repository_url?: string | null
  branch?: string | null
  scan_path?: string | null
  status: string
  findings_count: number
  critical_count: number
  high_count: number
  medium_count: number
  low_count: number
  policy_violation_count: number
  drift_count: number
  drift_detected: boolean
  ai_summary?: string | null
  created_at: string
}

export interface TerraformStats {
  total_workspaces: number
  total_resources: number
  total_findings: number
  open_findings: number
  critical_findings: number
  high_findings: number
  drift_count: number
  policy_violation_count: number
  severity_breakdown: Record<string, number>
  category_breakdown: Record<string, number>
}

export interface TerraformFindingListResponse {
  items: TerraformFinding[]
  total: number
  page: number
  page_size: number
}

export interface TerraformDriftListResponse {
  items: TerraformDrift[]
  total: number
  page: number
  page_size: number
}

export interface TerraformAnalysisResponse {
  workspace: TerraformWorkspace
  scan: TerraformScan
  findings: TerraformFinding[]
  drift: TerraformDrift[]
  stats: TerraformStats
}

export interface TerraformAnalyzeRequest {
  workspace_id?: string
  workspace_name?: string
  files?: Record<string, string>
  terraform_content?: string
  state?: Record<string, unknown>
  demo?: boolean
  scan_name?: string
}

interface TerraformListParams {
  page?: number
  page_size?: number
  workspace_id?: string
  severity?: string
  category?: string
  status?: string
}

export const terraformApi = {
  stats: () => api.get<TerraformStats>('/terraform/stats').then((r) => r.data),

  workspaces: () => api.get<TerraformWorkspace[]>('/terraform/workspaces').then((r) => r.data),

  scans: () => api.get<TerraformScan[]>('/terraform/scans').then((r) => r.data),

  findings: (params?: TerraformListParams) =>
    api.get<TerraformFindingListResponse>('/terraform/findings', { params }).then((r) => r.data),

  finding: (id: string) => api.get<TerraformFinding>(`/terraform/findings/${id}`).then((r) => r.data),

  drift: (params?: TerraformListParams) =>
    api.get<TerraformDriftListResponse>('/terraform/drift', { params }).then((r) => r.data),

  analyze: (payload: TerraformAnalyzeRequest) =>
    api.post<TerraformAnalysisResponse>('/terraform/analyze', payload).then((r) => r.data),
}

File: backend/app/schemas/terraform.py
"""Terraform security and drift schemas."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class TerraformWorkspaceCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=255)
    description: str | None = None
    source_type: str = "upload"
    repository_url: str | None = None
    branch: str | None = "main"
    root_path: str | None = None
    provider: str | None = None
    environment: str | None = None
    owner: str | None = None
    metadata: dict[str, Any] | None = None


class TerraformUploadRequest(TerraformWorkspaceCreate):
    files: dict[str, str] | None = None
    terraform_content: str | None = None
    state: dict[str, Any] | None = None


class TerraformAnalyzeRequest(BaseModel):
    workspace_id: str | None = None
    workspace_name: str | None = None
    files: dict[str, str] | None = None
    terraform_content: str | None = None
    state: dict[str, Any] | None = None
    demo: bool = False
    scan_name: str | None = None


class TerraformWorkspaceOut(BaseModel):
    id: str
    name: str
    description: str | None
    source_type: str
    repository_url: str | None
    branch: str | None
    root_path: str | None
    provider: str | None
    environment: str | None
    owner: str | None
    last_scan_id: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TerraformResourceOut(BaseModel):
    id: str
    workspace_id: str
    scan_id: str | None
    address: str
    type: str
    name: str
    provider: str | None
    module: str | None
    file_path: str | None
    line_number: int | None
    desired_config: dict[str, Any] | None
    actual_state: dict[str, Any] | None
    drift_status: str
    risk_score: float | None
    created_at: datetime

    model_config = {"from_attributes": True}


class TerraformFindingOut(BaseModel):
    id: str
    workspace_id: str
    scan_id: str | None
    resource_id: str | None
    title: str
    description: str
    impact: str | None
    severity: str
    category: str
    status: str
    scanner: str
    rule_id: str | None
    resource_address: str | None
    resource_type: str | None
    file_path: str | None
    line_number: int | None
    remediation: str | None
    ai_explanation: str | None
    confidence_score: float | None
    created_at: datetime

    model_config = {"from_attributes": True}


class TerraformDriftOut(BaseModel):
    id: str
    workspace_id: str
    scan_id: str | None
    resource_id: str | None
    resource_address: str
    resource_type: str | None
    attribute_path: str
    desired_value: Any | None
    actual_value: Any | None
    drift_type: str
    severity: str
    status: str
    description: str
    remediation: str | None
    confidence_score: float | None
    created_at: datetime

    model_config = {"from_attributes": True}


class TerraformPolicyViolationOut(BaseModel):
    id: str
    workspace_id: str
    scan_id: str | None
    finding_id: str | None
    policy_name: str
    rule_id: str | None
    message: str
    severity: str
    resource_address: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class TerraformScanOut(BaseModel):
    id: str
    workspace_id: str | None
    scan_name: str
    source_type: str
    repository_url: str | None
    branch: str | None
    scan_path: str | None
    status: str
    findings_count: int
    critical_count: int
    high_count: int
    medium_count: int
    low_count: int
    policy_violation_count: int
    drift_count: int
    drift_detected: bool
    ai_summary: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class TerraformDashboardStats(BaseModel):
    total_workspaces: int
    total_resources: int
    total_findings: int
    open_findings: int
    critical_findings: int
    high_findings: int
    drift_count: int
    policy_violation_count: int
    severity_breakdown: dict[str, int]
    category_breakdown: dict[str, int]


class TerraformFindingListResponse(BaseModel):
    items: list[TerraformFindingOut]
    total: int
    page: int
    page_size: int


class TerraformDriftListResponse(BaseModel):
    items: list[TerraformDriftOut]
    total: int
    page: int
    page_size: int


class TerraformAnalysisResponse(BaseModel):
    workspace: TerraformWorkspaceOut
    scan: TerraformScanOut
    resources: list[TerraformResourceOut]
    findings: list[TerraformFindingOut]
    drift: list[TerraformDriftOut]
    policy_violations: list[TerraformPolicyViolationOut]
    stats: TerraformDashboardStats
